lib/api/ldp/ldp-request.js

```diff
@@ -0,0 +1,145 @@
+'use strict'
+
+const HttpError = require('standard-http-error')
+
+class LdpRequest {
+  /**
+   * @param options {object}
+   *
+   * @param options.host {SolidHost} Host / config object
+   *
+   * @param options.target {URL} Parsed full request url
+   *
+   * @param options.req {HttpRequest} Save the Express req object, useful
+   *   to access headers/preferences, writeStream for body, etc.
+   *
+   * @param [options.authn=null] {object} Authentication object, contains WebID
+   *   string, as well as any bearer credentials/tokens. Needed for
+   *   authenticated fetch (of remote group ACLs, of Copy resources, etc).
+   *   Null if request is not authenticated.
+   *
+   * @param [options.parsedBody] {?} Request body, parsed when appropriate
+   *   (when content-type is parse-able)
+   */
+  constructor (options) {
+    this.host = options.host
+    this.target = options.target
+    this.req = options.req
+    this.authn = options.authn
+    this.parsedBody = options.parsedBody
+  }
+
+  static from ({req, host}) {
+    let request
+
+    switch (req.method) {
+      case 'head':
+        /**
+         * Checks to see if target exists
+         */
+        request = LdpHeadRequest.from({req, host})
+        break
+      case 'get':
+        /**
+         * Use cases to handle:
+         *  - LdpListContainerRequest, if target is an ldp container
+         *  - However, if Accept: header is HTML, check if index.html exists
+         *  - Need to also support similar use case if index.ttl exists
+         *  - Otherwise, plain LdpGetRequest
+         */
+        request = LdpGetRequest.from({req, host})
+        break
+      case 'put':
+        /**
+         * Creates a resource or container, and any necessary containers above it in
+         * the hierarchy. Idempotent.
+         *
+         * If target is a container:
+         *   - creates the container if none existed
+         *   - does nothing if container exists, does not delete/clear existing contents
+         *   - has `mkdir -p` semantics (creates missing container hierarchy)
+         *
+         * If target is a resource:
+         *   - writes the resource, always overwriting existing contents if any existed
+         *     (since we don't support conditional `If-None-Match` requests)
+         *   - has `mkdir -p` semantics (creates missing container hierarchy)
+         */
+        request = LdpPutRequest.from({req, host})
+        break
+      case 'post':
```

Comment:
Currently, the target path for POST is required to be a container. If it's not a container, solid-server currently throws a 405 Method Not Allowed error.

Should we continue to implement this behavior? I'm thinking yes, since POST to a resource doesn't seem to make much sense.

My other question is - should the "must be a container" requirement be signaled by a 405 error, or would it be more appropriate to return a 400 error instead?
(If the former, we need to make sure we return the Allow header per spec - "The response MUST include an Allow header containing a list of valid methods for the requested resource.")

Reply:
Seems like a 405 case to me, that's the most specific one.